File: cpp_nn/initializers.cpp
#include "initializers.h"


float_batch Constant::initialize(unsigned int w, unsigned int h)
{
    float_batch temp(w, vector<float>(h, 1));

    for(unsigned int i  = 0; i < w; i++){
        for(unsigned int j = 0; j < h; j++){
            temp[i][j] = this->c;
        }
    }
    return temp;
}

float_batch RandomUniform::initialize(unsigned int w, unsigned int h){
    float_batch temp(w, vector<float>(h, 1));
    std::mt19937 gen(this->rd());
    for(unsigned int i  = 0; i < w; i++){
        for(unsigned int j = 0; j < h; j++){
            temp[i][j] = this->dis(gen);
        }
    }
    return temp;
}

File: python_nn/pure_nn/initializers.py
from abc import ABC

import random


def supported_initializers():
    return [x.__name__ for x in Initializer.__subclasses__()]


class Initializer:
    def initialize(self, x):
        raise NotImplementedError


class Constant(Initializer, ABC):
    def __init__(self, c=0):
        self._c = c

    def initialize(self, x):
        if isinstance(x[0], int):
            w = len(x)
            temp = [None for _ in range(w)]
            for i in range(w):
                temp[i] = self._c
            return temp
        elif isinstance(x[0], list):
            w, h = len(x), len(x[0])
            temp = [[None for _ in range(h)] for _ in range(w)]
            for i in range(w):
                for j in range(h):
                    temp[i][j] = self._c
            return temp
        else:
            raise TypeError


class RandomUniform(Initializer, ABC):
    def initialize(self, x):
        w, h = len(x), len(x[0])
        temp = [[None for _ in range(h)] for _ in range(w)]
        for i in range(w):
            for j in range(h):
                temp[i][j] = random.uniform(0, 1)
        return temp
